cmd/snap/cmd_run.go

```diff
@@ -216,6 +220,74 @@ func snapRunHook(snapName, snapRevision, hookName string) error {
 	return runSnapConfine(info, hook.SecurityTag(), snapName, "", hook.Name, nil)
 }
 
+func migrateXauthority(info *snap.Info) error {
+	u, err := userCurrent()
+	if err != nil {
+		return fmt.Errorf(i18n.G("cannot get the current user: %v"), err)
+	}
+
+	// If we're running as root then we don't do anything
+	if u.Uid == "0" {
+		return nil
+	}
+
+	xauthPath := getEnv("XAUTHORITY")
+	if len(xauthPath) == 0 || !osutil.FileExists(xauthPath) {
+		// Nothing to do for us. Most likely running outside of any
+		// graphical X11 session.
+		return nil
+	}
+
+	// Copy Xauthority file into a temporary place so we can safely
+	// process it further there.
+	tmpDir, err := ioutil.TempDir("", "xauth")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+	flags := osutil.CopyFlagSync|osutil.CopyFlagPreserveAll|osutil.CopyFlagOverwrite
+	tmpXauthPath := filepath.Join(tmpDir, "Xauthority")
+	err = osutil.CopyFile(xauthPath, tmpXauthPath, flags)
+	if err != nil {
+		return err
+	}
+
+	baseTargetDir := filepath.Join(dirs.XdgRuntimeDirBase, u.Uid)
+	targetPath := filepath.Join(baseTargetDir, "Xauthority")
+
+	// Only validate Xauthority file again when both files don't match
+	// ohterwise we can continue using the existing Xauthority file
+	if osutil.FileExists(targetPath) && osutil.FilesAreEqual(targetPath, tmpXauthPath) {
+		setEnv("XAUTHORITY", targetPath)
+		return nil
+	}
+
+	cookies, err := x11.ValidateXauthority(tmpXauthPath)
+	if err != nil {
+		return err
+	} else if cookies == 0 {
+		// Don't return an error when the Xauthority file doesn't have
+		// any cookies. We simply do nothing and leave the XAUTHORITY
+		// environment variable as it is.
+		return nil
+	}
+
+	if !osutil.FileExists(baseTargetDir) {
+		os.MkdirAll(baseTargetDir, 0700)
+	}
+
+	err = osutil.CopyFile(tmpXauthPath, targetPath, flags)
```

Comment:
This can be done in a single line: `if err := osutil.CopyFile(...); err != nil {`.